pkg/providers/metrics/prom_check.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/kubevela/workflow/pkg/types"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	// ProviderName is provider name for install.
+	ProviderName = "metrics"
+)
+
+type provider struct{}
+
+// PromCheck do health check from metrics from prometheus
+func (h *provider) PromCheck(ctx monitorContext.Context, wfCtx wfContext.Context, v *value.Value, act types.Action) error {
+	stepID, err := v.GetString("stepID")
+	if err != nil {
+		return err
+	}
+
+	valueStr, err := getQueryResult(ctx, v)
+	if err != nil {
+		return err
+	}
+
+	conditionStr, err := v.GetString("condition")
+	if err != nil {
+		return err
+	}
+
+	res, err := compareValueWithCondition(valueStr, conditionStr, v)
+
+	if err != nil {
+		return err
+	}
+
+	if res {
+		// not meet the condition
+		return handleSuccessCompare(wfCtx, stepID, v, conditionStr, valueStr)
+	}
+	return handleFailCompare(wfCtx, stepID, v, conditionStr, valueStr)
+}
+
+func handleSuccessCompare(wfCtx wfContext.Context, stepID string, v *value.Value, conditionStr, valueStr string) error {
+	d, err := v.GetString("duration")
+	if err != nil {
+		return err
+	}
+	duration, err := time.ParseDuration(d)
+	if err != nil {
+		return err
+	}
+
+	st := getMetricsStatusTime(wfCtx, stepID, "success")
+	if st == 0 {
+		// first success
+		v.FillObject(fmt.Sprintf("The healthy condition should be %s, and the query result is %s, indicating success.", conditionStr, valueStr), "message")
+		setMetricsStatusTime(wfCtx, stepID, "success", time.Now().Unix())
+		setMetricsStatusTime(wfCtx, stepID, "fail", time.Now().Unix())
+		return v.FillObject(false, "result")
+	}
+	successTime := time.Unix(st, 0)
+	if successTime.Add(duration).Before(time.Now()) {
+		if err = v.FillObject("The metric check has passed successfully.", "message"); err != nil {
+			return err
+		}
+		return v.FillObject(true, "result")
+	}
+	v.FillObject(fmt.Sprintf("The healthy condition should be %s, and the query result is %s, indicating success. The success has persisted for %s, with success duration being %s.", conditionStr, valueStr, time.Since(successTime).String(), duration), "message")
+	return v.FillObject(false, "result")
+}
+
+func handleFailCompare(wfCtx wfContext.Context, stepID string, v *value.Value, conditionStr, valueStr string) error {
+	setMetricsStatusTime(wfCtx, stepID, "success", 0)
+	ft := getMetricsStatusTime(wfCtx, stepID, "")
+	d, err := v.GetString("failDuration")
+	if err != nil {
+		return err
+	}
+	failDuration, err := time.ParseDuration(d)
+	if err != nil {
+		return err
+	}
+
+	if ft == 0 {
+		// first failed
+		setMetricsStatusTime(wfCtx, stepID, "fail", time.Now().Unix())
+		v.FillObject(fmt.Sprintf("The healthy condition should be %s, but the query result is %s, indicating failure, with the failure duration being %s. The check has terminated.", conditionStr, valueStr, failDuration), "message")
```

Comment:
Maybe we can wrap two `fillObject` in success and fail check into a function for reusing.

Reply:
Since the message are different between different failed cases, do we really need to merge them and parse the message as a parameter of function?